k8s: Resource[T], an implementation of informers with per-sub queues

joamaki opened the pull request:
The Resource[T] provides access to a Kubernetes resource as an
observable stream of events or as via a read-only Store[T]:

```
type Resource[T k8sRuntime.Object] interface {
    stream.Observable[Event[T]]
    Store(context.Context) (Store[T], error)
}
```

An event is either *SyncEvent, *UpdateEvent or *DeleteEvent. The
event can be processed synchronously with event.Handle(onSync,
onUpdate, onDelete), or asynchronously with a type switch and call
to event.Done(err). On non-nil errors from event handling the key
corresponding to the object being handled is requeued for later
processing and after retries are exhausted the stream completes with
the error from final retry.

The Store() method blocks until the resource has been synchronized with
the api-server (e.g. CachesSynced == true).  Store[T] is a wrapper around
cache.Store that only provides the methods that are safe to use (Get,
GetByKey, List, ListKeys) and that are specialized to type T.

This abstracts the management of an informer and cache behind the
`NewResourceConstructor` function that takes a cache.ListerWatcher
and provides a constructor that can be used from a cell to provide the
Resource to the application:

```
var exampleCell = hive.NewCell(
       "example",
       fx.Provide(
               // Provide `Resource[*slim_corev1.Pod]` to the application:
               resource.NewResourceConstructor(func(c k8sClient.Clientset) cache.ListerWatcher {
                       return utils.ListerWatcherFromTyped[*slim_corev1.PodList](c.Slim().CoreV1().Pods(""))
               }),
               // Use the resource:
               newExample,
       }),
)

func newExample(pods resource.Resource[*slim_corev1.Pod]) *Example {
           e := &Example{...}
           pods.Observe(e.ctx, e.onPodEvent, e.onPodsComplete)
           return e
}
func (e *Example) onPodEvent(key resource.Key, ev resource.Event[*slim_core.Pod]) error {
           // Process event ...
}
func (e *Example) onPodsComplete(err error) {
           // Handle error. We may end up here when:
           // 1. Observing context cancelled (e.g. we're stopping)
           // 2. Resource is being stopped
           // 3. Processing an event has failed after maximum number of retries.
}
```

If multiple resources are being used, it may make sense to go via channels to ensure
sequential processing and to avoid locks:

```
type Foo struct {
    ctx context.Context
    pods resource.Resource[*slim_corev1.Pod]
    services resource.Resource[*slim_corev1.Service]
}

func (f *Foo) processLoop() {
  errs := make(chan error)
  pods := stream.ToChannel(f.ctx, errs, f.pods)
  services := stream.ToChannel(f.ctx, errs, f.services)
  for {
    select {
    case ev := <-pods:
      ev.Handle(f.podsSynced, f.podUpdated, f.podsComplete)
    case ev := <-services:
      ev.Handle(f.servicesSynced, f.serviceUpdated, f.servicesComplete)
    case err := <-errs:
      // handle error by e.g. full state reset and replay etc.
    }
  }
}
```

joamaki committed:
logging: Collapse constructors by module in PrintObjects()

This significantly simplifies the output for cells that export lots
of objects. We're not that interested in the actual constructor name
anyway as that can be easily found.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
k8s: Add resource package

The Resource[T] provides access to a Kubernetes resource as an
observable stream of events or as via a read-only Store[T]:

   type Resource[T k8sRuntime.Object] interface {
   	stream.Observable[Event[T]]
   	Store(context.Context) (Store[T], error)
   }

An event is either *SyncEvent, *UpdateEvent or *DeleteEvent. The
event can be processed synchronously with event.Handle(onSync,
onUpdate, onDelete), or asynchronously with a type switch and call
to event.Done(err). On non-nil errors from event handling the key
corresponding to the object being handled is requeued for later
processing and after retries are exhausted the stream completes with
the error from final retry.

The Store() method blocks until the resource has been synchronized with
the api-server (e.g. CachesSynced == true).  Store[T] is a wrapper around
cache.Store that only provides the methods that are safe to use (Get,
GetByKey, List, ListKeys) and that are specialized to type T.

This abstracts the management of an informer and cache behind the
`NewResourceConstructor` function that takes a cache.ListerWatcher
and provides a constructor that can be used from a cell to provide the
Resource to the application:

   var exampleCell = hive.NewCell(
           "example",
           fx.Provide(
                   // Provide `Resource[*slim_corev1.Pod]` to the application:
                   resource.NewResourceConstructor(func(c k8sClient.Clientset) cache.ListerWatcher {
                           return utils.ListerWatcherFromTyped[*slim_corev1.PodList](c.Slim().CoreV1().Pods(""))
                   }),
           }),
           ...
   )

Resource is implemented as an informer whose handlers enqueue the
changes keys to per-subscriber queues, from which a key is dequeued,
the objects then retrieved from the cache.Store and then provided
to the subscriber. This allows subscribers to process the changes
at their own pace and allows coalescing changes to the same key when
the subscriber is slow.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki commented on pkg/k8s/resource/resource.go:
```diff
@@ -254,12 +251,57 @@ func (r *resource[T]) Observe(subCtx context.Context, next func(Event[T]), compl
+		// Blocks until rate limiter says it is OK to continue.
+		kq.AddRateLimited(key)
+
+		// TODO: log error (or should the subscriber log as it has more context?) + metrics.
```
Thoughts on this?

joamaki commented on pkg/k8s/resource/resource.go:
```diff
@@ -0,0 +1,307 @@
+
+		// Can still retry processing it. Add it back to the queue.
+		// Blocks until rate limiter says it is OK to continue.
+		kq.AddRateLimited(key)
```
TODO: This should probably be called in the background as we don't want to block other work from being processed?

squeed commented on pkg/k8s/resource/resource.go:
```diff
@@ -0,0 +1,350 @@
+		queue.Add(&syncEntry{})
+		r.mu.Unlock()
+
+		for {
```
Does it make sense to plumb in parallel workers here?

joamaki replied:
I would keep that separate, e.g. if a subscriber wants to process the events in parallel, then it can e.g. do `stream.ToChannel` and have a worker pool feed from it. This works fine as the queue entry is marked processed via `Event.Done()` so it can be done completely asynchronously.

squeed commented on pkg/k8s/resource/event.go:
```diff
@@ -0,0 +1,80 @@
+	b.done(err)
+}
+
+// SyncEvent is emitted when the store has completed the initial synchronization
```
What does this mean, exactly? Does it indicate simply that the store is loaded? Does it mean that every pre-existing item has been Observed at least once?

joamaki replied:
This means that the set of updated objects preceding the sync event constitute a complete snapshot and that the store can now be used (e.g. ListKeys() will not return a partially synced view). 

In practical terms we need the sync event for garbage collection purposes. For example when the agent starts we would synchronize state to BPF maps and then run a garbage collection to see if there are entries in the BPF maps that no longer have a corresponding k8s object.

I'll try to improve the documentation!

joamaki committed:
stream: Don't close errs channel in ToChannel

ToChannel is often used with multiple observables in a for-select loop
with a shared errs channel. The channel should be created and closed by
the caller.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
k8s: Add example for Resource[T]

Example for how Resource[T] can be used in a complete application.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

dylandreimerink reviewed (approved):
LGTM
